### ISSUE TITLE
Evaluate the best approach for translation of dynamic UI labels

### ISSUE THREAD

**Issue by chiatt** [archesproject/arches#6956]

**Comment by apeters** [comment]
There are 3 potential paths to localizing the dynamic content within Arches and I've set up 3 branches in Arches that users can test.  The branches only attempt to localize card name.  You should be able to save and view card names in multiple languages.

1. Using the package django-translated-fields
    **Branch:** https://github.com/archesproject/arches/tree/L10n_TranslatedField_package
    **Description:**
        This package works by wrapping the field you're interested in localizing with a TranslatedField method.

        name = TranslatedField(
            models.TextField(blank=True, null=True),
            attrgetter=fallback_to_default
        )

    **Pros:**
        Simple to setup
        Minimal edits required to work
        Virtually no impact on client-side code

    **Cons:**
        Adds columns to the tables for every language and field that needs to be localized (eg: name above, becomes name_en, name_es, name_de, etc...)
        Export/Import of localized graphs could be very complicated
        Need to run/preserve a local migration every time you add/remove a language
        Would not work for localized business data

    **Verdict:**
        All the cons listed above make me uncomfortable.  
        Having schemas change based on the languages of interest seems problematic especially when sharing graphs with others


2. Using a single json object per model field to hold all loclized versions of a string
    **Branch:** https://github.com/archesproject/arches/tree/L10n_JSONField
    **Description:**
        Uses a JSONField to replace an existing TextField
        Data is stored as JSON keyed to each language 
        EG:


        {
                "en": "Today",
                "de": "Heute",
                "es": "Hoy"
        }

        
     This json object is then passed directly to the client
        
    **Pros:**
        Minimal edits to server side (python) code 
        Json object is passed directly to the ui 
        UI has complete awareness of all languages for a given string

    **Cons:**
        Requires a lot of edits to client-side code
        Would potentially break any custom project templates, views, scripts, etc...
        Would not work for localized business data

    **Verdict:**
        The fact that custom project code could break (which is usually out of our contol), seems a big red flag.


3. Using a hybrid approach that combines the best of the 2 solutions above
    **Branch:** https://github.com/archesproject/arches/tree/L10n_Hybrid
    **Description:**
        Uses a custom TranlatedJSONField to replace an existing TextField
        Data is stored as JSON keyed to each language 
        EG:

            {
                "en": "Today",
                "de": "Heute",
                "es": "Hoy"
            }

      The TranlatedJSONField returns a custom L10n_Field which is a class that can return 
        a single language based on the active language of the request (the default), or it can return the entire json object 
        
    **Pros:**
        Minimal edits to server side (python) code 
        Virtually no impact on client-side code
        Graphs with localized cards, widgets, etc. can easily be shared

    **Cons:**
        Would not work for localized business data

    **Verdict:**
        This seems like the best option.  Client-side code isn't disrupted and hence any custom client-side project code is also safe.
        Languages can be added dynamically without having to run a migration and there are no impacts to the schema beyond the initial work.

**Comment by whatisgalen** [comment]
for approaches 2 and 3, why wouldn't storing as json work for localized business data? Wouldn't the mapping file contain something indicating what language to store/read a string from/into?

**Comment by apeters** [comment]
Storing business data as JSON definitely will work, but localizing the entire tile.data object (as I've done with card.name) is not the path forward.  Instead we will just need to localize the string datatype (so just a portion of the data in tile.data).  Doing that means making a complex object out of what used to be just a string.

**Comment by azaroth42** [comment]
To throw a cat among the pigeons ... what about text direction? It is not necessarily determined by language, nor is it able to be determined by unicode code points.   I recommend this: https://www.w3.org/TR/string-meta/ and (then) https://w3c.github.io/rdf-dir-literal/#introduction

TLDR: A simple hash of language/value is insufficient in the general case, and instead it needs an array of language/direction/value data structures

**Comment by apeters** [comment]
I think as far as graph metadata is concerned (card and widget labels, etc..) text direction can be reasonably determined from the language tag (or maybe an as yet to be created setting in the system).  Now as far as business data goes, I think there is definitely a case where users might like to select the language and direction of the text being entered separate from the selection of the language of the site they are viewing the data in.

**Comment by apeters** [comment]
At this point I think we've settled on using the Hybrid approach from above.
